test/unit/test_vectorisation.py

```diff
@@ -0,0 +1,100 @@
+#     * Redistributions in binary form must reproduce the above copyright
+#       notice, this list of conditions and the following disclaimer in the
+#       documentation and/or other materials provided with the distribution.
+#     * The name of Imperial College London or that of other
+#       contributors may not be used to endorse or promote products
+#       derived from this software without specific prior written
+#       permission.
+#
+# THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTERS
+# ''AS IS'' AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT
+# LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS
+# FOR A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE
+# COPYRIGHT HOLDERS OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT,
+# INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES
+# (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR
+# SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION)
+# HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT,
+# STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
+# ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED
+# OF THE POSSIBILITY OF SUCH DAMAGE.
+import numpy as np
+from pyop2 import op2
+from pyop2.configuration import configuration
+import glob
+import os
+import pytest
+
+
+some_vectorisation_keys = ["__attribute__", "vector_size", "aligned", "#pragma omp simd"]
+
+
+class TestVectorisation:
+
+    @pytest.fixture
+    def s(self):
+        return op2.Set(1)
+
+    @pytest.fixture
+    def md1(self, s):
+        n = op2.Dat(s, [3], np.float64)
+        o = op2.Dat(s, [4], np.float64)
+        md = op2.MixedDat([n, o])
+        return md
+
+    @pytest.fixture
+    def md(self, s):
+        n = op2.Dat(s, [4], np.float64)
+        o = op2.Dat(s, [5], np.float64)
+        md = op2.MixedDat([n, o])
+        return md
+
+    def test_vectorisation(self, md1, md):
+        # Test that vectorised code produced the correct result
+        ret = md.inner(md1)
+        assert abs(ret - 32) < 1e-12
+        ret = md1.inner(md)
+        assert abs(ret - 32) < 1e-12
+
+        # Test that we actually vectorised
+        list_of_files = glob.glob(configuration["cache_dir"]+"/*/*.c")
+        latest_file = max(list_of_files, key=os.path.getctime)
+        with open(latest_file, 'r') as file:
+            generated_code = file.read()
+        assert (all(key in generated_code for key in some_vectorisation_keys)), "The kernel for an inner product has not been succesfully vectorised."
+
+    def test_no_vectorisation(self, md1, md):
+        # turn vectorisation off
+        op2.init(**{"vectorization_strategy": ""})
+
+        # Test that unvectorised code produced the correct result
+        ret = md.inner(md1)
+        assert abs(ret - 32) < 1e-12
+        ret = md1.inner(md)
+        assert abs(ret - 32) < 1e-12
+
+        # Test that we did not vectorise
```

Comment:
Also is there a better way to get the file to the generated code? This probably breaks if we run it in parallel.

Reply:
I feel like a better way to check the generated code would be to create a `Parloop` and then inspect `parloop.global_kernel.code_to_compile` ([link](https://github.com/OP2/PyOP2/blob/master/pyop2/parloop.py#L147), [link](https://github.com/OP2/PyOP2/blob/master/pyop2/global_kernel.py#L325)). You can look in `test/unit/test_indirect_loop.py` for an example but then you should use `LegacyParloop` instead of `par_loop` to get the right object back.

Reply:
In `test/unit/test_indirect_loop.py` all Kernels are generated from a string not from loopy. I didn't want to write a loopy kernel myself so I took a test where we are automatically generating a loopy kernel which we can vectorise already anyways. But I guess you are right because I have no access to the kernel that is he doing the calculations for `inner`

Reply:
Oh yes you're quite right. I think it would still be worth writing a loopy kernel so we can do this check without having to look at the filesystem.

Reply:
Actually I just unrolled now whatever was happening in the `inner` def on the dat, so I got around writing the kernel myself :)